Kernel/KBO.cpp

```diff
@@ -567,72 +554,71 @@ KBO KBO::testKBO()
 #if __KBO__CUSTOM_PREDICATE_WEIGHTS__
       KboWeightMap<PredSigTraits>::randomized(), 
 #endif
-      funcPrec(),
-      predPrec(),
+      DArray<int>::fromIterator(getRangeIterator(0, (int)env.signature->functions())),
+      DArray<int>::fromIterator(getRangeIterator(0, (int)env.signature->typeCons())),
+      DArray<int>::fromIterator(getRangeIterator(0, (int)env.signature->predicates())),
       predLevels(),
       false);
 }
 
-void KBO::zeroOutWeightForMaximalFuncs() {
+void KBO::zeroWeightForMaximalFunc() {
   // actually, it's non-constant maximal func, as constants cannot be weight 0
 
-  using SortType = TermList;
   using FunctionSymbol = unsigned;
   auto nFunctions = _funcWeights._weights.size();
-  auto maximalFunctions = Map<SortType, FunctionSymbol>();
 
-  for (FunctionSymbol i = 0; i < nFunctions; i++) {
-    auto symb = env.signature->getFunction(i);
-    auto sort = symb->fnType()->result();
-    auto maxFn = maximalFunctions.getOrInit(sort, [&](){ return i; } );
+  FunctionSymbol maxFn = 0;
+
+  for (FunctionSymbol i = 1; i < nFunctions; i++) {
     if (compareFunctionPrecedences(maxFn, i) == LESS) {
-      maximalFunctions.replace(sort, i);
+      maxFn = i;
     }
   }
 
-  Map<SortType, FunctionSymbol>::Iterator it(maximalFunctions);
-  while (it.hasNext()) {
-    FunctionSymbol i = it.next().value();
-    auto symb = env.signature->getFunction(i);
-    auto arity = symb->arity();
+  auto symb = env.signature->getFunction(maxFn);
+  auto arity = symb->numTermArguments();
 
-    // skip constants here (they can't be smaller than $var)
-    if (arity == 0) continue;
+  // skip constants here (they can't be smaller than $var)
+  if (arity != 0){
     // TODO: we could also have remembered the "second largest" symbol, if a constant was the largest
     // (but we could not use them if they were of arity 1)
 
-    _funcWeights._weights[i] = 0;
+    _funcWeights._weights[maxFn] = 0;
   }
+  
 }
 
 template<class HandleError>
 void KBO::checkAdmissibility(HandleError handle) const 
 {
-  using SortType = TermList;
   using FunctionSymbol = unsigned;
   auto nFunctions = _funcWeights._weights.size();
-  auto maximalFunctions = Map<SortType, FunctionSymbol>();
+  FunctionSymbol maxFn = 0;
 
-  for (FunctionSymbol i = 0; i < nFunctions; i++) {
-    auto sort = env.signature->getFunction(i)->fnType()->result();
-    /* register min function */
-    auto maxFn = maximalFunctions.getOrInit(sort, [&](){ return i; } );
+  for (FunctionSymbol i = 1; i < nFunctions; i++) {
     if (compareFunctionPrecedences(maxFn, i) == LESS) {
-      maximalFunctions.replace(sort, i);
+      maxFn = i;
     }
   }
 
+  // TODO remove unary minus check once new 
+  // theory calculus goes into master
+  auto isUnaryMinus = [](unsigned functor){
```

Comment:
I locally made some changes so that the unary minus doesn't get added to the signature here, if it is not present in the problem. I'm not sure whether i should push it to this branch or make a separate PR. What do you think?
It's a quite small thing.

Reply:
Don't have a preference either way. Adding it to this PR is fine by me.

Reply:
Yes, @joe-hauns, please add here if you haven't already. Thank you!